core/src/main/scala/org/apache/spark/sql/delta/DeltaOptions.scala

```diff
@@ -90,6 +91,15 @@ trait DeltaWriteOptionsImpl extends DeltaOptionParser {
     options.get(DATA_CHANGE_OPTION).exists(!toBoolean(_, DATA_CHANGE_OPTION))
   }
 
+  /** Whether to only overwrite partitions that have data written into it at runtime. */
+  def dynamicPartitionOverwriteMode: Boolean = {
+    val mode = options.get(PARTITION_OVERWRITE_MODE)
+      .map(_.toUpperCase(Locale.ROOT))
+      .getOrElse(sqlConf.getConf(SQLConf.PARTITION_OVERWRITE_MODE))
+    require(mode == "STATIC" || mode == "DYNAMIC",
```

Comment:
how will incorrect modes get handles nicely? i dont want to silently turn a typo like say `dynamc` into static mode, which could lead to data loss. i realize this is an imperfect place to check it but i have no better place to do so.

Reply:
My thinking is to assume `static`. Anything that's not `dynamic` leads to `static` incl. typos. Whatever needs this option could `logInfo` to "announce" what's used. I think it'd make the code clearer.

Reply:
static is the one where you delete all your data. it makes the code clearer but the reality of a typo a lot scarier!

Reply:
Makes sense. Why not use [TypedConfigBuilder.checkValues](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala#L114-L123) then?

Reply:
Just today I found an example how to use `checkValues` (in the [codebase of Delta Lake](https://github.com/delta-io/delta/blob/e36dc6b9ca8ea8e893080dcea847978d5835125b/core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala#L190-L196)):

```
  val DELTA_PROTOCOL_DEFAULT_WRITER_VERSION =
    buildConf("properties.defaults.minWriterVersion")
      .doc("The default writer protocol version to create new tables with, unless a feature " +
        "that requires a higher version for correctness is enabled.")
      .intConf
      .checkValues(Set(1, 2, 3, 4))
      .createWithDefault(2)
```

Reply:
interesting! 
example of how it is used is here:
```
    val readerVersion = configs.get(MIN_READER_VERSION_PROP.toLowerCase(Locale.ROOT))                                                                                   
      .map(getVersion(MIN_READER_VERSION_PROP, _))                                                                                                                      
      .getOrElse(conf.getConf(DeltaSQLConf.DELTA_PROTOCOL_DEFAULT_READER_VERSION))    
```
however it seems checkValues only applies to what is set in `conf` (which is `spark.sessionState.conf`), doesnt it?
i dont think this would catch an invalid version set in `configs`?
there is so some basic checking in `getVersion`, which raises an IllegalArgumentException if needed.

Reply:
No idea. Would have to investigate myself further. We can skip this for now if that's the only blocker.

Reply:
> Another major point i completely missed in this first pass. According to spark docs, the SQL conf should make INSERT OVERWRITE follow dynamic partitioning too. But I dont see any tests with INSERT OVERWRITE. I think that must be tested as well.

ok i went down the rabbit hole of trying to understand how this works. seems like `DeltaTableV2` supports `V1_BATCH_WRITE` (and uses `InsertableRelation`) but not `BATCH_WRITE`. my limited understanding here is that `V1_BATCH_WRITE` does not support `OVERWRITE_DYNAMIC`? is that correct?